### DISCUSSION TITLE
Maintain URL Parameter Order

### DISCUSSION THREAD

**Discussion by tariwiencke** [47ng/nuqs#777] [Ideas] [answered]
It would be beneficial for Nuqs to include a feature that ensures URL query parameters are kept in a consistent order. When adding or modifying query parameters via Nuqs utilities, the order of parameters in the URL should remain consistent and predictable. The order can be based on the useQueryStates object order.

**Comment by franky47** [comment]
There was a proposal to sort search params alphabetically in #638, but after playing with it, it felt very unpredictable. See https://github.com/47ng/nuqs/pull/638#issuecomment-2352471673

The current behaviour is in line with URLSearchParams: new keys are appended at the end of the query string, and updates to existing values are done in-place.

Doing it such as it follows the order of the object raises the question of what happens when combining multiple useQueryStates or useQueryState updates in one batch, example:

```ts
const [, set1] = useQueryStates({ a: ..., b: ... })
const [, set2] = useQueryStates({ c: ..., d: ... })
const [, set3] = useQueryState('e')

set3('e')
set2({ c: 'c' })
set1({ b: 'b' })
```

**Comment by franky47** [comment, marked as answer]
This is available in `nuqs@2.6.0-beta.2`, feel free to try it:

Complete docs (with demo): https://nuqs-next.47ng.com/docs/options#processing-urlsearchparams

```tsx
<NuqsAdapter
  processUrlSearchParams={(search) => {
    const entries = Array.from(search.entries())
    entries.sort(([a], [b]) => a.localeCompare(b))
    return new URLSearchParams(entries)
  }}
>
  {children}
</NuqsAdapter>
```
